pkg/products/marin3r/apiUsagePrometheusRules.go

```diff
@@ -41,59 +41,84 @@ func (r *Reconciler) newAlertsReconciler(grafanaConsoleURL string) (resources.Al
-					Annotations: map[string]string{
-						"message": fmt.Sprintf(
-							"RHOAM API usage is between %s and %s of the allowable threshold, %d requests per %s, during the last %s",
-							alertConfig.MinRate, upperMessage, rateLimitRequestsPerUnit, rateLimitUnit, alertConfig.Period,
-						),
-						"grafanaConsole": string(grafanaConsoleURL),
+		switch alertConfig.Type {
+		case marin3rconfig.AlertTypeSpike:
+
+			expr := fmt.Sprintf(
+				"max_over_time((increase(ratelimit_service_rate_limit_apicast_ratelimit_generic_key_slowpath_total_hits[1m]))[%s:]) /2 > %d",
+				alertConfig.Period, rateLimitRequestsPerUnit)
+
+			alert := resources.AlertConfiguration{
+				AlertName: "rate-limit-spike",
+				GroupName: "ratelimit-spike.rules",
+				Namespace: namespace,
+				Interval:  alertConfig.Period,
+				Rules: []monitoringv1.Rule{
+					{
+						Alert: alertConfig.RuleName,
+						Annotations: map[string]string{
+							"message":        fmt.Sprintf("hard limit of %d breached at least once in the last %s", rateLimitRequestsPerUnit, alertConfig.Period),
+							"grafanaConsole": grafanaDashboardURL,
+						},
+						Expr:   intstr.FromString(expr),
+						Labels: map[string]string{"severity": alertConfig.Level},
+					},
+				},
+			}
+			result = append(result, alert)
+		case marin3rconfig.AlertTypeThreshold:
+			usageFrequencyMins, err := intervalToMinutes(alertConfig.Period)
+			if err != nil {
+				return nil, err
+			}
+			requestsAllowedOverTimePeriod := requestsAllowedPerSecond * float64(usageFrequencyMins*60)
+
+			minRateValue, maxRateValue, err := parsePercenteageRange(
+				alertConfig.MinRate,
+				alertConfig.MaxRate,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			lowerExpr := increaseExpr(totalRequestsMetric, alertConfig.Period, ">=", requestsAllowedOverTimePeriod, &minRateValue)
+			upperExpr := increaseExpr(totalRequestsMetric, alertConfig.Period, "<=", requestsAllowedOverTimePeriod, maxRateValue)
+
+			// Get the complete expression by ANDing the lower and the upper if the
+			// upper limit is set, if not, assign the lower one
+			expr := *lowerExpr
+			upperMessage := "*"
+			if upperExpr != nil {
+				expr = fmt.Sprintf("%s and %s", expr, *upperExpr)
+				upperMessage = *alertConfig.MaxRate
+			}
+
+			alert := resources.AlertConfiguration{
+				AlertName: alertName,
+				GroupName: "api-usage.rules",
+				Namespace: namespace,
+				Rules: []monitoringv1.Rule{
+					{
+						Alert: alertConfig.RuleName,
+						Annotations: map[string]string{
+							"message": fmt.Sprintf(
+								"Total API usage in your API Management service is between %s%% and %s%% of the allowable threshold, %d requests per %s, during the last %s",
+								alertConfig.MinRate, upperMessage, rateLimitRequestsPerUnit, rateLimitUnit, alertConfig.Period,
+							),
+							"grafanaConsole": grafanaDashboardURL,
+						},
+						Expr:   intstr.FromString(expr),
+						Labels: map[string]string{"severity": alertConfig.Level},
```

Comment:
```suggestion
                var alert resources.AlertConfiguration
		switch alertConfig.Type {
		case marin3rconfig.AlertTypeSpike:
		        alert = mapSpikeAlert(alertConfig /*, other parameters */)
                case marin3rconfig.AlertTypeThreshold:
                        alert = mapThresholdAlert(alertConfig /*, other parameters */)
                default:
                        alert = mapThresholdAlert(alertConfig /*, other parameters */)
```

Maybe we could externalize the creation of each type of alert in functions? It'd make the switch shorter, and we could reuse the threshold alert function to default in case the type doesn't match (existing config maps that don't have this field or invalid values)

Reply:
Yes, think it would be easier to read also. 👍

Reply:
Do we definitely want to do a default where it's the ThresholdAlert? I would expect if it's not either of those types we don't want to do anything as it may be the case that a new type is added and the switch is maybe just not updated?

Reply:
Makes sense, another option would be to `continue` on default, maybe log that an unsupported type was found. I was thinking of defaulting  to this in case the operator is run on a cluster that already has the ConfigMap (which wouldn't include the new field), but will probably occur just in development so not really a concern

Reply:
I've added an info log the default. Agree it would be good to at least log so that the it's obvious in the future if there is a case where the type is unsupported. 👍